### PR TITLE
Magnifier for iOS 17+

### PR DESCRIPTION
**Pull request by alexzhirkevich**
## Proposed Changes

- Text field and selection container magnifier for iOS 17+

## Testing

Test: try magnifier in all types of text fields

## Issues Fixed

Partly: https://github.com/JetBrains/compose-multiplatform/issues/3692

### CODE REVIEW

**Comment by alexzhirkevich** [comment]
What is the correct way to get local `UIView` from foundation module? Seems like there no more `LocalLayerContainer` and interop container is something different

**Comment by mazunin-v-jb** [comment]
@alexzhirkevich afaik for now we don't have such API
@dima-avdeev-jb correct me please if I'm wrong

**Comment by mazunin-v-jb** [comment]
PR looks good to me. Although, it shouldn't go to the next release, so I'll leave this PR without an approval for now.
And we should wait for the second review.

**Comment by alexzhirkevich** [comment]
There is 1 issue. 
Text selection in debug build causes significant FPS drops on my old 12 mini (especially in low power mode, but in default mode too). There are no drops when displaying magnifier outside of text bounds (for example in empty text field). I measured time of magnifier position calculation and it's fine: most of the time it's <1ms and sometimes it's 1ms (in LPM at the high cpu load moments).

Then I checked `jb-main` branch and it has such FPS drops too. Magnifier just makes it much more noticeable because of freezing spring animation. 

Can it potentially be some expensive calls in selection logic?

Here is what i'm talking about (jb-main and magnifier):



https://github.com/JetBrains/compose-multiplatform-core/assets/63979218/bbe8d2d8-4704-4953-bd52-21bd3b46ab34


https://github.com/JetBrains/compose-multiplatform-core/assets/63979218/8226737e-1b44-4d04-9b82-b7da952a97d4

**Comment by MatkovIvan** [comment]
@alexzhirkevich Sorry for the delay, the team is busy with the upcoming 1.6.0 release. We'll come back to this after that

**Comment by mazunin-v-jb** [comment]
@alexzhirkevich 
We're ready to review. Could you please resolve merge conflicts?

**Comment by alexzhirkevich** [comment]
Fixed conflicts, but found a bug. 
When you select some text and delete it or start typing while holding the loupe - it getting stuck. Appreciate any help

**Comment by alexzhirkevich** [comment]
> When you select some text and delete it or start typing while holding the loupe - it getting stuck

Fixed it. The problem was when `SelectionHandle` with `detectDragGestures` modifier is disposed before drag finish neither `onDragEnd` nor `onDragCancel` is called. I guess it wasn't reproducible on android only because you can't use software keyboard while selecting text (at least on android device i have).

File: compose/foundation/foundation/src/skikoMain/kotlin/androidx/compose/foundation/text/CupertinoTextFieldPointerModifier.skiko.kt
Comment (ASalavei) [inline compose/foundation/foundation/src/skikoMain/kotlin/androidx/compose/foundation/text/CupertinoTextFieldPointerModifier.skiko.kt:179]: Redundant comma at the end of the line.

Comment (MatkovIvan) [reply compose/foundation/foundation/src/skikoMain/kotlin/androidx/compose/foundation/text/CupertinoTextFieldPointerModifier.skiko.kt:179]: It's OK, we usually do that to avoid git diff/conflicts when adding more parameters or array elements

File: compose/foundation/foundation/src/skikoMain/kotlin/androidx/compose/foundation/text/CupertinoTextFieldPointerModifier.skiko.kt
Comment (ASalavei) [inline compose/foundation/foundation/src/skikoMain/kotlin/androidx/compose/foundation/text/CupertinoTextFieldPointerModifier.skiko.kt]: Could you please remove comments like `// required for magnifier` ?

File: compose/foundation/foundation/src/uikitMain/kotlin/androidx/compose/foundation/Magnifier.uikit.kt
Comment (ASalavei) [inline compose/foundation/foundation/src/uikitMain/kotlin/androidx/compose/foundation/Magnifier.uikit.kt]: Please also remove comment.

File: compose/foundation/foundation/src/uikitMain/kotlin/androidx/compose/foundation/text/selection/SelectionManager.uikit.kt
Comment (ASalavei) [inline compose/foundation/foundation/src/uikitMain/kotlin/androidx/compose/foundation/text/selection/SelectionManager.uikit.kt:48]: Should `calculateSelectionMagnifierCenterIOS`  also be called be here?

**Comment by ASalavei** [comment]
Just to warn: because of the platform layers refactoring done [here](https://github.com/JetBrains/compose-multiplatform-core/pull/1199), it might be a problem with magnifier positioning. Depending on which MR is merged first, the bug may need to be fixed in another one.

File: compose/foundation/foundation/src/uikitMain/kotlin/androidx/compose/foundation/text/selection/SelectionManager.uikit.kt
Comment (alexzhirkevich) [reply compose/foundation/foundation/src/uikitMain/kotlin/androidx/compose/foundation/text/selection/SelectionManager.uikit.kt:48]: There are 2 incompatible `calculateSelectionMagnifierCenterAndroid` implementations for `TextFieldSelectionManager` and `SelectionManager`. As for `SelectionContainer` iOS magnifier behavior is almost the same as on Android so i just left android implementation here.

File: compose/foundation/foundation/src/uikitMain/kotlin/androidx/compose/foundation/text/selection/TextFieldSelectionManager.uikit.kt
Comment (ASalavei) [inline compose/foundation/foundation/src/uikitMain/kotlin/androidx/compose/foundation/text/selection/TextFieldSelectionManager.uikit.kt]: It seems we can remove TODO here because of compose text selection logic you described below.

**Comment by ASalavei** [comment]
@MatkovIvan , do you have rights to grant access for report generator in this MR?
https://github.com/JetBrains/compose-multiplatform-core/actions/runs/8402704969/job/23059307784?pr=1000

File: compose/foundation/foundation/src/uikitMain/kotlin/androidx/compose/foundation/Magnifier.uikit.kt
Comment (alexzhirkevich) [inline compose/foundation/foundation/src/uikitMain/kotlin/androidx/compose/foundation/Magnifier.uikit.kt]:
```suggestion
```

File: compose/foundation/foundation/src/uikitMain/kotlin/androidx/compose/foundation/text/selection/TextFieldSelectionManager.uikit.kt
Comment (alexzhirkevich) [inline compose/foundation/foundation/src/uikitMain/kotlin/androidx/compose/foundation/text/selection/TextFieldSelectionManager.uikit.kt]:
```suggestion
// native magnifier also hides when selection goes to the next line in multiline text field
```

**Review by ASalavei — APPROVED** [review]
LGTM from my side. Thank you for the feature!

**Review by mazunin-v-jb — APPROVED** [review]
looks good to me too
thank you for PR!

**Comment by elijah-semyonov** [comment]
The PR is related to a crash on pre iOS 17 devices
https://github.com/JetBrains/compose-multiplatform/issues/4644

The symbol [UITextLoupeSession](https://developer.apple.com/documentation/uikit/uitextloupesession) is uniquely referenced here.

File: compose/foundation/foundation/src/uikitMain/kotlin/androidx/compose/foundation/Magnifier.uikit.kt
Comment (alexzhirkevich) [inline compose/foundation/foundation/src/uikitMain/kotlin/androidx/compose/foundation/Magnifier.uikit.kt:350]: There is a check for iOS version. It is implemented incorrectly?

Comment (elijah-semyonov) [reply compose/foundation/foundation/src/uikitMain/kotlin/androidx/compose/foundation/Magnifier.uikit.kt:350]: It's implemented correctly, but the problem is due to linking. K/N can see the symbol and tried to link it during the launch.

Comment (elijah-semyonov) [reply compose/foundation/foundation/src/uikitMain/kotlin/androidx/compose/foundation/Magnifier.uikit.kt:350]: I'll promote this problem to K/N team, the only known workaround as of now is to use strings and C-apis directly to avoid mentioning the symbol altogether.

Comment (alexzhirkevich) [reply compose/foundation/foundation/src/uikitMain/kotlin/androidx/compose/foundation/Magnifier.uikit.kt:350]: >use strings and C-apis directly

Do you mean using cinterop like [here ](https://github.com/JetBrains/compose-multiplatform-core/tree/jb-main/compose/ui/ui-uikit/src/uikitMain/objc/CMPUIKitUtils) or something different?

Comment (elijah-semyonov) [reply compose/foundation/foundation/src/uikitMain/kotlin/androidx/compose/foundation/Magnifier.uikit.kt:350]: No, I actually meant barebone ObjC runtime, but @ASalavei came with [a much better solution](https://github.com/JetBrains/compose-multiplatform-core/pull/1278#pullrequestreview-2006273682) to simply hide the symbols from cinterop

Comment (alexzhirkevich) [reply compose/foundation/foundation/src/uikitMain/kotlin/androidx/compose/foundation/Magnifier.uikit.kt:350]: > simply hide the symbols from cinterop

I thought about it too. Nice solution. 
I seen some places in skiko where iOS 13 is checked the same way and class/method is loaded. Should it be fixed as well?

Comment (elijah-semyonov) [reply compose/foundation/foundation/src/uikitMain/kotlin/androidx/compose/foundation/Magnifier.uikit.kt:350]: Methods are not a big deal, ObjC will simply throw unrecognised selector exception in case we don't do a version check. The classes are a problem (currently), can you leave a link?

Comment (ASalavei) [reply compose/foundation/foundation/src/uikitMain/kotlin/androidx/compose/foundation/Magnifier.uikit.kt:350]: I found one place in Skiko, and it seems like No, there are no iOS 13 specific classes, only a method call, which is wrapped into proper check.

**Pull request merged**
